src/Microsoft.ML.Tokenizers/Model/SentencePieceBaseModel.cs

```diff
@@ -0,0 +1,751 @@
+                        EncodeByte(enumerator.Current, OneByteUtf8EncodingMaxId, ByteCodeToIdOffset, ref bytesCount, ref bytesPoolArray, ref sb);
+                    }
+                }
+                else
+                {
+                    if (bytesCount >= 1)
+                    {
+                        FlushBytes(ref bytesCount, ref bytesPoolArray, ref charPoolArray, ref sb);
+                    }
+
+                    if (!TryDecodeAsSpecialToken(this, enumerator.Current, considerSpecialTokens, ref sb) && TryMapIdToToken(enumerator.Current, out string? token))
+                    {
+                        AppendTokenWithCheckingPrefix(AddDummyPrefix, TreatWhitespaceAsSuffix, token!, prefixSuffixChar, ref sb, ref prefixRemoved, ref suffixIndex);
+                    }
+                }
+            }
+
+            if (bytesCount >= 1)
+            {
+                FlushBytes(ref bytesCount, ref bytesPoolArray, ref charPoolArray, ref sb);
+            }
+
+            if (AddDummyPrefix && TreatWhitespaceAsSuffix && suffixIndex >= 0 && sb.Length > 0)
+            {
+                Debug.Assert(sb[suffixIndex] == SentencePieceNormalizer.DummyPrefix);
+                Debug.Assert(sb.Length > suffixIndex);
+
+                sb.Remove(suffixIndex, 1);
+            }
+
+            if (bytesPoolArray is not null)
+            {
+                ArrayPool<byte>.Shared.Return(bytesPoolArray);
+            }
+
+            if (charPoolArray is not null)
+            {
+                ArrayPool<char>.Shared.Return(charPoolArray);
+            }
+
+            return EscapeWhiteSpaces ? sb.ToString(SentencePieceNormalizer.DummyPrefix, ' ') : sb.ToString();
+
+            static void FlushBytes(ref int bytesCount, ref byte[]? bytesPoolArray, ref char[]? charPoolArray, ref ValueStringBuilder sb)
+            {
+                Debug.Assert(bytesCount >= 1);
+                Debug.Assert(bytesPoolArray is not null);
+
+                int len = Encoding.UTF8.GetMaxCharCount(bytesCount);
+
+                charPoolArray ??= ArrayPool<char>.Shared.Rent(Math.Max(len, 50));
+
+                if (len > charPoolArray.Length)
+                {
+                    Helpers.ArrayPoolGrow(ref charPoolArray, len);
+                }
+
+                int charCount = Helpers.GetChars(bytesPoolArray.AsSpan(0, bytesCount), charPoolArray);
+
+                sb.Append(charPoolArray.AsSpan(0, charCount));
+                bytesCount = -1;
+            }
+
+            static void EncodeByte(int id, int oneByteUtf8EncodingMaxId, int byteCodeToIdOffset, ref int bytesCount, ref byte[]? bytesPoolArray, ref ValueStringBuilder sb)
+            {
+                if (id <= oneByteUtf8EncodingMaxId)
+                {
+                    sb.Append((char)(id - byteCodeToIdOffset));
+                }
+                else
+                {
+                    bytesCount = 1;
+                    bytesPoolArray ??= ArrayPool<byte>.Shared.Rent(50);
```

Comment:
I see this 50 value in a few places. Where does it come from? Can we put it in a named const?